File: jobify/shared/ats_detect.py
"""jobify.shared.ats_detect — Detect which ATS platform a job URL uses (M-3).

Routing layer used by tailor's prepare flow and (eventually) submit's
deterministic flow. Each known ATS gets its own per-platform DOM-based
handler (Ashby, Greenhouse, Lever) that reads from ``job["form_answers"]``
and never makes Anthropic API calls. Anything we don't have a handler
for falls through to ``UniversalApplicant``, which drives the page via a
prepare-only Claude tool-use agent (M-4 strips its submit-mode tool).

The ``USE_LEGACY_APPLICANTS`` env flag is gone — per-ATS handlers are the
default and the cheap path. The vision agent is the fallback for the
messier middle of the funnel (Workday, iCIMS, SmartRecruiters,
aggregator-direct postings).

Moved from ``jobify/tailor/applicant/detector.py`` in PR-4 because the
routing decision is shared infrastructure, not tailor-specific. The
``applicant/detector.py`` path remains as a re-export shim until PR-9.
"""

from __future__ import annotations

import logging

logger = logging.getLogger("jobify.shared.ats_detect")


def detect_ats(url: str) -> str:
    """Identify the ATS platform from the application URL.

    Returns one of: ashby, greenhouse, lever, indeed, workday, linkedin,
    icims, smartrecruiters, generic.
    """
    url_lower = (url or "").lower()

    if "ashby" in url_lower or "ashbyhq.com" in url_lower or "ashby_jid" in url_lower:
        return "ashby"
    if (
        "greenhouse.io" in url_lower
        or "boards.greenhouse" in url_lower
        or "job-boards.greenhouse" in url_lower
    ):
        return "greenhouse"
    if "lever.co" in url_lower or "jobs.lever" in url_lower:
        return "lever"
    if "indeed.com/applystart" in url_lower or "indeed.com/viewjob" in url_lower:
        return "indeed"
    if "myworkdayjobs.com" in url_lower or "workday.com" in url_lower:
        return "workday"
    if "linkedin.com" in url_lower:
        return "linkedin"
    if "icims.com" in url_lower:
        return "icims"
    if "smartrecruiters.com" in url_lower:
        return "smartrecruiters"
    return "generic"


def get_applicant(url: str):
    """Return the appropriate applicant instance for a given URL.

    Per-ATS DOM handlers (zero Anthropic spend) are the default for
    Ashby, Greenhouse, and Lever. Everything else falls through to
    ``UniversalApplicant``, which uses a Claude tool-use agent in
    prepare-only mode (M-4) to drive unknown ATSes by vision.

    PR-7 removed the legacy ``_bootstrap_tailor_sys_path()`` stub that this
    function used to call before importing the prepare_dom adapters. The
    moved adapters in ``jobify.submit.adapters.prepare_dom.*`` and their
    sibling ``prepare_loop`` now use explicit jobify-namespaced imports
    for everything (``BaseApplicant``, ``BrowserSession``, ``url_resolver``,
    ``config``, ``prompts``), so the sys.path side-effect is no longer
    needed to make those modules importable.
    """
    ats = detect_ats(url)
    logger.info(f"ats detected: {ats} for {url}")

    if ats == "ashby":
        from jobify.submit.adapters.prepare_dom.ashby import AshbyApplicant
        return AshbyApplicant()
    if ats == "greenhouse":
        from jobify.submit.adapters.prepare_dom.greenhouse import GreenhouseApplicant
        return GreenhouseApplicant()
    if ats == "lever":
        from jobify.submit.adapters.prepare_dom.lever import LeverApplicant
        return LeverApplicant()

    # Fallback for Workday / iCIMS / SmartRecruiters / Indeed / aggregators.
    # The vision agent is prepare-only — it can never click Submit (M-4).
    from jobify.submit.adapters.prepare_dom.universal import UniversalApplicant
    return UniversalApplicant()

File: web/lib/submit/atsDetect.ts
// web/lib/submit/atsDetect.ts — TS port of the ATS URL-detection heuristics.
//
// Ports + merges (union of substring patterns per ATS):
//   - jobify/shared/ats_detect.py::detect_ats            (broad baseline)
//   - jobify/submit/adapters/prepare_dom/greenhouse.py::GreenhouseApplicant.detect
//   - jobify/submit/adapters/prepare_dom/lever.py::LeverApplicant.detect
//   - jobify/submit/adapters/prepare_dom/ashby.py::AshbyApplicant.detect
//
// Only the 8 kinds the pinned `SubmitPacket.posting.ats_kind` contract cares
// about — no `indeed` branch (the Python baseline has one; this TS port
// intentionally narrows to what Task 5 needs, so an Indeed URL falls through
// to "generic"). Case-insensitive substring matching, same defensive style as
// the Python originals: `(url || "").toLowerCase()` before any comparison.
// Pure function, no I/O — this only classifies whatever URL string it's given;
// aggregator URL *resolution* is a later, out-of-scope phase.

export type AtsKind =
  | "greenhouse"
  | "lever"
  | "ashby"
  | "workday"
  | "icims"
  | "smartrecruiters"
  | "linkedin"
  | "generic";

export function detectAtsKind(url: string | null | undefined): AtsKind {
  const urlLower = (url || "").toLowerCase();

  // Ashby — union of ats_detect.py ("ashby", "ashbyhq.com", "ashby_jid") and
  // ashby.py ("ashbyhq.com", "ashby_jid", "jobs.ashby").
  if (
    urlLower.includes("ashby") ||
    urlLower.includes("ashbyhq.com") ||
    urlLower.includes("ashby_jid") ||
    urlLower.includes("jobs.ashby")
  ) {
    return "ashby";
  }

  // Greenhouse — union of ats_detect.py ("greenhouse.io", "boards.greenhouse",
  // "job-boards.greenhouse") and greenhouse.py ("boards.greenhouse.io",
  // "job-boards.greenhouse.io", "apply.greenhouse.io",
  // "greenhouse.io/embed/job_app").
  if (
    urlLower.includes("greenhouse.io") ||
    urlLower.includes("boards.greenhouse") ||
    urlLower.includes("job-boards.greenhouse") ||
    urlLower.includes("apply.greenhouse.io") ||
    urlLower.includes("greenhouse.io/embed/job_app")
  ) {
    return "greenhouse";
  }

  // Lever — union of ats_detect.py ("lever.co", "jobs.lever") and lever.py
  // ("jobs.lever.co", "jobs.eu.lever.co").
  if (
    urlLower.includes("lever.co") ||
    urlLower.includes("jobs.lever") ||
    urlLower.includes("jobs.eu.lever")
  ) {
    return "lever";
  }

  // Workday — ats_detect.py only (no per-ATS adapter exists yet).
  if (urlLower.includes("myworkdayjobs.com") || urlLower.includes("workday.com")) {
    return "workday";
  }

  // LinkedIn — ats_detect.py only.
  if (urlLower.includes("linkedin.com")) {
    return "linkedin";
  }

  // iCIMS — ats_detect.py only.
  if (urlLower.includes("icims.com")) {
    return "icims";
  }

  // SmartRecruiters — ats_detect.py only.
  if (urlLower.includes("smartrecruiters.com")) {
    return "smartrecruiters";
  }

  return "generic";
}
